threadpoolctl.py

```diff
@@ -599,6 +211,520 @@ def get_original_num_threads(self):
+                user_api = _ALL_USER_APIS
+            elif user_api in _ALL_USER_APIS:
+                user_api = [user_api]
+            else:
+                raise ValueError(
+                    "user_api must be either in {} or None. Got "
+                    "{} instead.".format(_ALL_USER_APIS, user_api))
+
+            if limits is not None:
+                limits = {api: limits for api in user_api}
+            prefixes = []
+        else:
+            if isinstance(limits, list):
+                # This should be a list of dicts of modules, for compatibility
+                # with the result from threadpool_info.
+                limits = {module["prefix"]: module["num_threads"]
+                          for module in limits}
+            elif isinstance(limits, _ThreadpoolInfo):
+                # To set the limits from the modules of a _ThreadpoolInfo
+                # object.
+                limits = {module.prefix: module.num_threads
+                          for module in limits}
+
+            if not isinstance(limits, dict):
+                raise TypeError("limits must either be an int, a list or a "
+                                "dict. Got {} instead".format(type(limits)))
+
+            # With a dictionary, can set both specific limit for given modules
+            # and global limit for user_api. Fetch each separately.
+            prefixes = [prefix for prefix in limits if prefix in _ALL_PREFIXES]
+            user_api = [api for api in limits if api in _ALL_USER_APIS]
+
+        return limits, user_api, prefixes
+
+    def _set_threadpool_limits(self):
+        """Change the maximal number of threads in selected thread pools.
+
+        Return a list with all the supported modules that have been found
+        matching `self._prefixes` and `self._user_api`.
+        """
+        if self._limits is None:
+            return None
+        else:
+            modules = _ThreadpoolInfo(prefixes=self._prefixes,
+                                      user_api=self._user_api)
+            for module in modules:
+                # self._limits is a dict {key: num_threads} where key is either
+                # a prefix or a user_api. If a module matches both, the limit
+                # corresponding to the prefix is chosed.
+                if module.prefix in self._limits:
+                    num_threads = self._limits[module.prefix]
+                else:
+                    num_threads = self._limits[module.user_api]
+
+                if num_threads is not None:
+                    module.set_num_threads(num_threads)
+            return modules
+
+
+@_format_docstring(
+    PREFIXES=", ".join('"{}"'.format(prefix) for prefix in _ALL_PREFIXES),
+    USER_APIS=", ".join('"{}"'.format(api) for api in _ALL_USER_APIS),
+    BLAS_LIBS=", ".join(_ALL_BLAS_LIRARIES),
+    OPENMP_LIBS=", ".join(_ALL_OPENMP_LIBRARIES))
+class _ThreadpoolInfo():
+    """Collection of all supported modules that have been found
+
```

Comment:
Maybe add a not to explain that the object oriented API of `_ThreadpoolInfo` and its model is private and the public API (i.e. the "threadpool_info" function) only exposes the "list of dicts" representation returned by the `.todicts` method.